pkg/ttnpb/contextual_end_device.go

```diff
@@ -0,0 +1,23 @@
+// Copyright © 2019 The Things Network Foundation, The Things Industries B.V.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package ttnpb
+
+import "context"
+
+// ContextualEndDevice is an end device with its context.
+type ContextualEndDevice struct {
```

Comment:
Since it's only used in JS, I think it's fine to put it there

Reply:
We get import cycles then; `pkg/joinserver/redis` can't import `pkg/joinserver`

Reply:
That should not be the case actually
`pkg/joinserver` should not import `pkg/joinserver/redis`, but `pkg/joinserver/redis` should be able to do the reverse.
In any case, `pkg/joinserver/internal` is also an option.